pkg/util/resource/resource_manager.go

```diff
@@ -0,0 +1,297 @@
+	}
+	if defaultExpiration <= 0 {
+		defaultExpiration = DefaultExpiration
+	}
+	if defaultTypeExpirations == nil {
+		defaultTypeExpirations = make(map[string]time.Duration)
+	}
+	m := &manager{
+		resources:              make(map[string]map[string]*item),
+		checkInterval:          checkInterval,
+		defaultExpiration:      defaultExpiration,
+		defaultTypeExpirations: defaultTypeExpirations,
+		stop:                   make(chan struct{}),
+	}
+	go m.backgroundGC()
+	return m
+}
+
+func (m *manager) backgroundGC() {
+	ticker := time.NewTicker(m.checkInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			m.gc()
+		case <-m.stop:
+			m.mu.Lock()
+			for _, typMap := range m.resources {
+				for _, item := range typMap {
+					item.res.Close()
+				}
+			}
+			m.resources = nil
+			m.mu.Unlock()
+			return
+		}
+	}
+}
+
+func (m *manager) gc() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	now := time.Now().UnixNano()
+	for typ, typMap := range m.resources {
+		for resName, item := range typMap {
+			select {
+			case lastTime := <-item.updateTimeChan:
+				if item.expiration >= 0 {
+					item.expiration = lastTime
+				}
+			case <-item.deleteMark:
+				item.res.Close()
+				delete(typMap, resName)
+			default:
+				if item.expiration >= 0 && item.expiration <= now {
+					item.res.Close()
+					delete(typMap, resName)
+				}
+			}
+		}
+		if len(typMap) == 0 {
+			delete(m.resources, typ)
+		}
+	}
+}
+
+func (m *manager) updateExpire(item *item) {
+	select {
+	case item.updateTimeChan <- time.Now().UnixNano() + item.res.KeepAliveTime().Nanoseconds():
+	default:
+	}
+}
+
+type NewResourceFunc func() (Resource, error)
+
+func (m *manager) Get(typ, name string, newResourceFunc NewResourceFunc) (Resource, error) {
+	m.mu.RLock()
+	typMap, ok := m.resources[typ]
+	if ok {
+		item := typMap[name]
+		if item != nil {
+			m.mu.RUnlock()
+			m.updateExpire(item)
+			return item.res, nil
+		}
+	}
+	m.mu.RUnlock()
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	typMap, ok = m.resources[typ]
+	if !ok {
+		typMap = make(map[string]*item)
+		m.resources[typ] = typMap
+	}
+	ite, ok := typMap[name]
+	if !ok {
+		res, err := newResourceFunc()
+		if err != nil {
+			return nil, err
+		}
+		if res.KeepAliveTime() == 0 {
+			defaultExpiration := m.defaultTypeExpirations[typ]
+			if defaultExpiration == 0 {
+				defaultExpiration = m.defaultExpiration
+			}
+			res = NewResource(WithResource(res), WithKeepAliveTime(defaultExpiration))
+		}
+		ite = &item{
+			res:            res,
+			updateTimeChan: make(chan int64, 1),
+			deleteMark:     make(chan struct{}, 1),
+		}
+		typMap[name] = ite
+	}
+	m.updateExpire(ite)
+	return ite.res, nil
+}
+
+func (m *manager) Delete(typ, name string) Resource {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	typMap, ok := m.resources[typ]
+	if !ok {
+		return nil
+	}
+	ite, ok := typMap[name]
+	if !ok {
+		return nil
+	}
+	select {
+	case ite.deleteMark <- struct{}{}:
+	default:
+	}
+	return ite.res
+}
+
+func (m *manager) Close() {
```

Comment:
It is recommended to include a mechanism that waits for the completion of the backgroundGC goroutine in this context.

Reply:
fixed